ir/src/main/scala/org/scalajs/core/ir/Printers.scala

```diff
@@ -1052,4 +1052,601 @@ object Printers {
     def complete(): Unit = ()
   }
 
+  abstract class RawIndentationManager extends IndentationManager {
+    private var indentSize = 0
+    private val indentString = "  "
+
+    override def indent(): Unit = indentSize += 1
+    override def undent(): Unit = indentSize -= 1
+
+    protected def print(s: String): Unit = {
+      out.write(indentString * indentSize)
+      out.write(s)
+    }
+
+    protected def println(s: String): Unit =
+      print(s + "\n")
+
+    override protected def println(): Unit =
+      out.write("\n")
+  }
+
+  class IRASTPrinter(protected val out: Writer) extends RawIndentationManager {
+    /** Pretty-prints the IR's AST nodes.
+      */
+    def printTopLevelTree(tree: Tree): Unit =
+      println(tree)
+
+    private final def println(tree: Tree): Unit = {
+      print(tree)
+      out.write("\n")
+    }
+
+    private final def printList(ts: List[Tree]): Unit = {
+      if (ts.isEmpty) {
+        print("List ()")
+      } else {
+        println("List (")
+        indent()
+        var rest = ts
+        while (rest.nonEmpty) {
+          print(rest.head)
+          rest = rest.tail
+          if (rest.nonEmpty)
+            println()
+        }
+        undent()
+        paren()
+      }
+    }
+
+    private final def printlnList(ts: List[Tree]): Unit = {
+      printList(ts)
+      println()
+    }
+
+    private final def printIdentList(is: List[Ident]): Unit = {
+      if (is.isEmpty) {
+        print("List ()")
+      } else {
+        println("List (")
+        indent()
+        var rest = is
+        while (rest.nonEmpty) {
+          print(rest.head)
+          rest = rest.tail
+          if (rest.nonEmpty)
+            println()
+        }
+        undent()
+        paren()
+      }
+    }
+
+    private final def printlnIdentList(is: List[Ident]): Unit = {
+      printIdentList(is)
+      println()
+    }
+
+    private final def print(ident: Ident): Unit =
+      print(s"Ident(${ident.name}, ${ident.originalName})")
+
+    private final def println(ident: Ident): Unit = {
+      print(ident)
+      out.write("\n")
+    }
+
+    private def print(tpe: Type): Unit =
+      print(tpe.toString)
+
+
+    private final def println(tp: Type): Unit = {
+      print(tp)
+      out.write("\n")
+    }
+
+    private final def paren(): Unit = {
+      out.write(")")
```

Comment:
This looks like over-abstraction to me. I suggest writing `out.write(")")` explicitly instead of calling `paren()` (btw, `paren` is not a good name because we've got no idea whether it's a opening or closing paren)